File: apps/api/test/contacts.test.ts
import { describe, expect, it } from "bun:test";
import { familyScope } from "../src/db/scoped";
import {
  addMember,
  api,
  createBaby,
  createFamily,
  createUser,
  db,
  rig,
  setPlan,
  signIn,
} from "./helpers";

describe("contact scoped helpers", () => {
  it("creates a contact with baby links and hydrates them back", async () => {
    const { family, baby } = await rig();
    const fam = familyScope(db(), family.id);

    const created = await fam.createContact({
      name: "Dr. Hansen",
      role: "doctor",
      icon: "doctor",
      phone: "+47 22 00 00 00",
      email: "hansen@legesenteret.no",
      website: "legesenteret.no",
      notes: "Fastlege",
      babyIds: [baby.id],
    });

    expect(created).not.toBeNull();
    expect(created!.name).toBe("Dr. Hansen");
    expect(created!.icon).toBe("doctor");
    expect(created!.babies).toEqual([{ id: baby.id, name: "Rig baby" }]);
  });

  it("treats zero baby links as a family-wide contact", async () => {
    const { family } = await rig();
    const fam = familyScope(db(), family.id);

    const created = await fam.createContact({ name: "Mormor", babyIds: [] });

    expect(created!.babies).toEqual([]);
    expect(created!.role).toBeNull();
  });

  it("shares one contact across several babies", async () => {
    const { family, baby } = await rig();
    const sibling = await createBaby(family.id, "Sibling");
    const fam = familyScope(db(), family.id);

    const shared = await fam.createContact({
      name: "Dr. Hansen",
      babyIds: [baby.id, sibling.id],
    });

    expect(shared!.babies.map((b) => b.name).sort()).toEqual([
      "Rig baby",
      "Sibling",
    ]);
  });

  it("lists only this family's contacts, by name", async () => {
    const { family } = await rig();
    const otherFamily = await createFamily("Other family");
    const fam = familyScope(db(), family.id);
    const otherFam = familyScope(db(), otherFamily.id);

    await fam.createContact({ name: "Zita", babyIds: [] });
    await fam.createContact({ name: "Anna", babyIds: [] });
    await otherFam.createContact({ name: "Not ours", babyIds: [] });

    const rows = await fam.listContacts();
    expect(rows.map((r) => r.name)).toEqual(["Anna", "Zita"]);
  });

  it("replaces the link set on update and leaves it alone when omitted", async () => {
    const { family, baby } = await rig();
    const sibling = await createBaby(family.id, "Sibling");
    const fam = familyScope(db(), family.id);
    const created = await fam.createContact({
      name: "Mormor",
      babyIds: [baby.id],
    });

    const relinked = await fam.updateContact(
      created!.id,
      {},
      { babyIds: [sibling.id] },
    );
    expect(relinked!.babies).toEqual([{ id: sibling.id, name: "Sibling" }]);

    const renamed = await fam.updateContact(
      created!.id,
      { name: "Farmor" },
      {},
    );
    expect(renamed!.name).toBe("Farmor");
    expect(renamed!.babies).toEqual([{ id: sibling.id, name: "Sibling" }]);

    const unlinked = await fam.updateContact(created!.id, {}, { babyIds: [] });
    expect(unlinked!.babies).toEqual([]);
  });

  it("refuses to update or delete another family's contact", async () => {
    const { family } = await rig();
    const otherFamily = await createFamily("Other family");
    const otherFam = familyScope(db(), otherFamily.id);
    const theirs = await otherFam.createContact({
      name: "Theirs",
      babyIds: [],
    });

    const fam = familyScope(db(), family.id);
    expect(await fam.getContact(theirs!.id)).toBeNull();
    expect(await fam.updateContact(theirs!.id, { name: "Hijacked" }, {})).toBe(
      null,
    );
    expect(await fam.deleteContact(theirs!.id)).toBe(false);
    expect((await otherFam.getContact(theirs!.id))!.name).toBe("Theirs");
  });

  it("drops link rows when the baby is deleted, keeping the contact", async () => {
    const { family, baby } = await rig();
    const fam = familyScope(db(), family.id);
    const created = await fam.createContact({
      name: "Dr. Hansen",
      babyIds: [baby.id],
    });

    await fam.deleteBaby(baby.id);

    const after = await fam.getContact(created!.id);
    expect(after).not.toBeNull();
    expect(after!.babies).toEqual([]);
  });
});

describe("contacts API", () => {
  it("requires premium to create, but never to read, edit or delete", async () => {
    const { family, cookie, baby } = await rig();

    const denied = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Dr. Hansen" },
    });
    expect(denied.status).toBe(402);
    expect(((await denied.json()) as { code: string }).code).toBe(
      "PLAN_REQUIRED",
    );

    await setPlan(family.id, "premium");
    const created = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Dr. Hansen", role: "doctor", babyIds: [baby.id] },
    });
    expect(created.status).toBe(201);
    const contact = (await created.json()) as {
      id: string;
      babies: { id: string; name: string }[];
    };
    expect(contact.babies).toEqual([{ id: baby.id, name: "Rig baby" }]);

    // Soft-lock: a downgrade keeps the address book fully usable.
    await setPlan(family.id, "free");

    const listed = await api("/api/contacts", { cookie });
    expect(listed.status).toBe(200);
    expect((await listed.json()) as unknown[]).toHaveLength(1);

    const edited = await api(`/api/contacts/${contact.id}`, {
      method: "PATCH",
      cookie,
      body: { phone: "+47 22 00 00 00" },
    });
    expect(edited.status).toBe(200);
    expect(((await edited.json()) as { phone: string }).phone).toBe(
      "+47 22 00 00 00",
    );

    const removed = await api(`/api/contacts/${contact.id}`, {
      method: "DELETE",
      cookie,
    });
    expect(removed.status).toBe(200);
  });

  it("rejects a baby id from another family", async () => {
    const { family, cookie } = await rig();
    await setPlan(family.id, "premium");
    const otherFamily = await createFamily("Other family");
    const theirBaby = await createBaby(otherFamily.id, "Their baby");

    const res = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Sneaky", babyIds: [theirBaby.id] },
    });
    expect(res.status).toBe(400);
    expect(((await res.json()) as { code: string }).code).toBe(
      "INVALID_REFERENCE",
    );
  });

  it("rejects a baby id from another family on update too", async () => {
    const { family, cookie } = await rig();
    await setPlan(family.id, "premium");
    const created = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Mormor" },
    });
    const { id } = (await created.json()) as { id: string };

    const otherFamily = await createFamily("Other family");
    const theirBaby = await createBaby(otherFamily.id, "Their baby");

    const res = await api(`/api/contacts/${id}`, {
      method: "PATCH",
      cookie,
      body: { babyIds: [theirBaby.id] },
    });
    expect(res.status).toBe(400);
    expect(((await res.json()) as { code: string }).code).toBe(
      "INVALID_REFERENCE",
    );
  });

  it("never leaks another family's contacts across the API", async () => {
    const { family, cookie } = await rig();
    await setPlan(family.id, "premium");
    await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Ours" },
    });

    const outsider = await createUser("Outsider");
    const otherFamily = await createFamily("Other family");
    await addMember(outsider.id, otherFamily.id, "admin");
    const otherCookie = await signIn(outsider.email);

    const listed = await api("/api/contacts", { cookie: otherCookie });
    expect(listed.status).toBe(200);
    expect(await listed.json()).toEqual([]);
  });

  it("dedupes repeated baby ids instead of failing the batch", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");

    const res = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Mormor", babyIds: [baby.id, baby.id] },
    });
    expect(res.status).toBe(201);
    expect(((await res.json()) as { babies: unknown[] }).babies).toHaveLength(
      1,
    );
  });

  it("validates email and 404s on an unknown id", async () => {
    const { family, cookie } = await rig();
    await setPlan(family.id, "premium");

    const bad = await api("/api/contacts", {
      method: "POST",
      cookie,
      body: { name: "Nope", email: "not-an-email" },
    });
    expect(bad.status).toBe(400);

    const missing = await api("/api/contacts/does-not-exist", {
      method: "PATCH",
      cookie,
      body: { name: "Ghost" },
    });
    expect(missing.status).toBe(404);
  });
});

File: apps/server-go/internal/api/contacts_test.go
package api_test

import (
	"net/http"
	"testing"

	"github.com/refsdal/pjokk/server/internal/testrig"
)

// -----------------------------------------------------------------------
// Contacts — ports apps/api/test/contacts.test.ts's "contact scoped
// helpers" and "contacts API" describe blocks. The TS predecessor
// soft-locked creation behind premium (402 PLAN_REQUIRED); this Go port
// removes that gate entirely (see internal/api/contacts.go's package doc
// comment), so the 402-vs-premium test is replaced by
// TestCreateContactIsFreeFullCRUD below.
// -----------------------------------------------------------------------

func TestCreateContactWithBabyLinkHydrates(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Dr. Hansen",
		"role":    "doctor",
		"icon":    "doctor",
		"phone":   "+47 22 00 00 00",
		"email":   "hansen@legesenteret.no",
		"website": "legesenteret.no",
		"notes":   "Fastlege",
		"babyIds": []string{babyID},
	})
	if res.Status != http.StatusCreated {
		t.Fatalf("status = %d, body %s", res.Status, res.Raw)
	}
	if res.JSON["name"] != "Dr. Hansen" {
		t.Errorf("name = %v, want %q", res.JSON["name"], "Dr. Hansen")
	}
	if res.JSON["icon"] != "doctor" {
		t.Errorf("icon = %v, want %q", res.JSON["icon"], "doctor")
	}
	babies, _ := res.JSON["babies"].([]any)
	if len(babies) != 1 || babies[0].(map[string]any)["id"] != babyID {
		t.Errorf("babies = %v, want one entry for %q", babies, babyID)
	}
}

func TestCreateContactZeroBabiesIsFamilyWide(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")

	res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{"name": "Mormor"})
	if res.Status != http.StatusCreated {
		t.Fatalf("status = %d, body %s", res.Status, res.Raw)
	}
	babies, _ := res.JSON["babies"].([]any)
	if len(babies) != 0 {
		t.Errorf("babies = %v, want empty (family-wide)", babies)
	}
	if res.JSON["role"] != nil {
		t.Errorf("role = %v, want nil", res.JSON["role"])
	}
}

func TestCreateContactSharedAcrossSeveralBabies(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")
	siblingID := a.NewBaby(familyID, "Sibling")

	res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Dr. Hansen",
		"babyIds": []string{babyID, siblingID},
	})
	if res.Status != http.StatusCreated {
		t.Fatalf("status = %d, body %s", res.Status, res.Raw)
	}
	babies, _ := res.JSON["babies"].([]any)
	if len(babies) != 2 {
		t.Errorf("babies = %v, want 2", babies)
	}
}

func TestListContactsByNameFamilyScoped(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	_, otherCookie := a.NewFamily("Other family", "stranger@example.com")

	for _, name := range []string{"Zita", "Anna"} {
		if res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{"name": name}); res.Status != http.StatusCreated {
			t.Fatalf("create %q status = %d, body %s", name, res.Status, res.Raw)
		}
	}
	if res := a.Do(http.MethodPost, "/api/contacts", otherCookie, map[string]any{"name": "Not ours"}); res.Status != http.StatusCreated {
		t.Fatalf("other family create status = %d, body %s", res.Status, res.Raw)
	}

	listed := a.DoArray(http.MethodGet, "/api/contacts", cookie, nil)
	if listed.Status != http.StatusOK {
		t.Fatalf("status = %d, body %s", listed.Status, listed.Raw)
	}
	if len(listed.JSON) != 2 {
		t.Fatalf("listed = %v, want 2 contacts", listed.JSON)
	}
	if listed.JSON[0].(map[string]any)["name"] != "Anna" || listed.JSON[1].(map[string]any)["name"] != "Zita" {
		t.Errorf("listed names = %v, want [Anna Zita]", listed.JSON)
	}
}

func TestUpdateContactReplacesLinkSetOmittedUntouched(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")
	siblingID := a.NewBaby(familyID, "Sibling")

	created := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Mormor",
		"babyIds": []string{babyID},
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("create status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	relinked := a.Do(http.MethodPatch, "/api/contacts/"+id, cookie, map[string]any{"babyIds": []string{siblingID}})
	if relinked.Status != http.StatusOK {
		t.Fatalf("relink status = %d, body %s", relinked.Status, relinked.Raw)
	}
	babies, _ := relinked.JSON["babies"].([]any)
	if len(babies) != 1 || babies[0].(map[string]any)["id"] != siblingID {
		t.Errorf("babies after relink = %v, want one entry for %q", babies, siblingID)
	}

	renamed := a.Do(http.MethodPatch, "/api/contacts/"+id, cookie, map[string]any{"name": "Farmor"})
	if renamed.Status != http.StatusOK {
		t.Fatalf("rename status = %d, body %s", renamed.Status, renamed.Raw)
	}
	if renamed.JSON["name"] != "Farmor" {
		t.Errorf("name = %v, want %q", renamed.JSON["name"], "Farmor")
	}
	babiesAfterRename, _ := renamed.JSON["babies"].([]any)
	if len(babiesAfterRename) != 1 || babiesAfterRename[0].(map[string]any)["id"] != siblingID {
		t.Errorf("babies after unrelated rename = %v, want still one entry for %q (untouched)", babiesAfterRename, siblingID)
	}

	unlinked := a.Do(http.MethodPatch, "/api/contacts/"+id, cookie, map[string]any{"babyIds": []string{}})
	if unlinked.Status != http.StatusOK {
		t.Fatalf("unlink status = %d, body %s", unlinked.Status, unlinked.Raw)
	}
	babiesAfterUnlink, _ := unlinked.JSON["babies"].([]any)
	if len(babiesAfterUnlink) != 0 {
		t.Errorf("babies after empty-array PATCH = %v, want empty", babiesAfterUnlink)
	}
}

func TestContactCrossFamilyUpdateDeleteIs404(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	created := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{"name": "Ours"})
	if created.Status != http.StatusCreated {
		t.Fatalf("create status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	_, otherCookie := a.NewFamily("Other family", "stranger@example.com")
	hijack := a.Do(http.MethodPatch, "/api/contacts/"+id, otherCookie, map[string]any{"name": "Hijacked"})
	if hijack.Status != http.StatusNotFound {
		t.Errorf("cross-family PATCH status = %d, want 404", hijack.Status)
	}
	del := a.Do(http.MethodDelete, "/api/contacts/"+id, otherCookie, nil)
	if del.Status != http.StatusNotFound {
		t.Errorf("cross-family DELETE status = %d, want 404", del.Status)
	}
	// Still ours — the refusal deleted nothing.
	listed := a.DoArray(http.MethodGet, "/api/contacts", cookie, nil)
	if len(listed.JSON) != 1 {
		t.Errorf("listed after refused cross-family delete = %v, want 1", listed.JSON)
	}
}

// TestCreateContactIsFreeFullCRUD replaces the TS predecessor's
// 402-then-premium test: creation needs no plan at all on this port, and
// read/edit/delete were always open.
func TestCreateContactIsFreeFullCRUD(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	created := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Dr. Hansen",
		"role":    "doctor",
		"babyIds": []string{babyID},
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("create status = %d, body %s, want 201 (free — no plan gate)", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)
	babies, _ := created.JSON["babies"].([]any)
	if len(babies) != 1 || babies[0].(map[string]any)["id"] != babyID {
		t.Errorf("babies = %v, want one entry for %q", babies, babyID)
	}

	listed := a.DoArray(http.MethodGet, "/api/contacts", cookie, nil)
	if listed.Status != http.StatusOK || len(listed.JSON) != 1 {
		t.Fatalf("list status = %d, len = %d, body %s", listed.Status, len(listed.JSON), listed.Raw)
	}

	edited := a.Do(http.MethodPatch, "/api/contacts/"+id, cookie, map[string]any{"phone": "+47 22 00 00 00"})
	if edited.Status != http.StatusOK {
		t.Fatalf("PATCH status = %d, body %s", edited.Status, edited.Raw)
	}
	if edited.JSON["phone"] != "+47 22 00 00 00" {
		t.Errorf("phone = %v, want %q", edited.JSON["phone"], "+47 22 00 00 00")
	}

	removed := a.Do(http.MethodDelete, "/api/contacts/"+id, cookie, nil)
	if removed.Status != http.StatusOK {
		t.Errorf("DELETE status = %d, want 200", removed.Status)
	}
}

func TestCreateContactRejectsForeignBabyId(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	familyB, _ := a.NewFamily("Other family", "other@example.com")
	theirBabyID := a.NewBaby(familyB, "Their baby")

	res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Sneaky",
		"babyIds": []string{theirBabyID},
	})
	if res.Status != http.StatusBadRequest {
		t.Fatalf("status = %d, body %s, want 400", res.Status, res.Raw)
	}
	if res.JSON["code"] != "INVALID_REFERENCE" {
		t.Errorf("code = %v, want INVALID_REFERENCE", res.JSON["code"])
	}
}

func TestUpdateContactRejectsForeignBabyId(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	created := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{"name": "Mormor"})
	if created.Status != http.StatusCreated {
		t.Fatalf("create status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	familyB, _ := a.NewFamily("Other family", "other@example.com")
	theirBabyID := a.NewBaby(familyB, "Their baby")

	res := a.Do(http.MethodPatch, "/api/contacts/"+id, cookie, map[string]any{"babyIds": []string{theirBabyID}})
	if res.Status != http.StatusBadRequest {
		t.Fatalf("status = %d, body %s, want 400", res.Status, res.Raw)
	}
	if res.JSON["code"] != "INVALID_REFERENCE" {
		t.Errorf("code = %v, want INVALID_REFERENCE", res.JSON["code"])
	}
}

func TestContactsNeverLeakAcrossFamilies(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")
	if res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{"name": "Ours"}); res.Status != http.StatusCreated {
		t.Fatalf("create status = %d, body %s", res.Status, res.Raw)
	}

	_, otherCookie := a.NewFamily("Other family", "outsider@example.com")
	listed := a.DoArray(http.MethodGet, "/api/contacts", otherCookie, nil)
	if listed.Status != http.StatusOK {
		t.Fatalf("status = %d, body %s", listed.Status, listed.Raw)
	}
	if len(listed.JSON) != 0 {
		t.Errorf("listed = %v, want empty (another family's contacts must not leak)", listed.JSON)
	}
}

func TestCreateContactDedupesRepeatedBabyIds(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	res := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":    "Mormor",
		"babyIds": []string{babyID, babyID},
	})
	if res.Status != http.StatusCreated {
		t.Fatalf("status = %d, body %s", res.Status, res.Raw)
	}
	babies, _ := res.JSON["babies"].([]any)
	if len(babies) != 1 {
		t.Errorf("babies = %v, want deduped to 1", babies)
	}
}

func TestCreateContactBadEmailIs400AndUnknownIdIs404(t *testing.T) {
	a := testrig.App(t)
	_, cookie := a.NewFamily("Hansen", "parent@example.com")

	bad := a.Do(http.MethodPost, "/api/contacts", cookie, map[string]any{
		"name":  "Nope",
		"email": "not-an-email",
	})
	if bad.Status != http.StatusBadRequest {
		t.Errorf("bad email status = %d, want 400", bad.Status)
	}

	missing := a.Do(http.MethodPatch, "/api/contacts/does-not-exist", cookie, map[string]any{"name": "Ghost"})
	if missing.Status != http.StatusNotFound {
		t.Errorf("unknown id PATCH status = %d, want 404", missing.Status)
	}
}
